paulialgebra: Improved evaluate_pauli_product

johanbluecreek opened the pull request:
evaluate_pauli_product had several limitations before:
1. Acting on arguments that did not have the as_coeff_mul() gave errors.
2. The functions purpose seems to be as a work around for the XFAIL/FIXME that
   still remains in the code. But did not produce a satisfying end result.
3. It should respect commutative rules.

This meant that, for example:

1.
In [1]: from sympy.physics.paulialgebra import Pauli, evaluate_pauli_product
In [2]: evaluate_pauli_product(1)
[---]
AttributeError: 'int' object has no attribute 'as_coeff_mul'

2.
In [3]: from sympy import I
In [4]: evaluate_pauli_product(I \* Pauli(1) \* Pauli(2) \* Pauli(1) \* Pauli(2))
Out[4]: -sigma3 \* sigma1 \* sigma2

3.
In [5]: from sympy import symbols
In [6]: tau1 = symbols("tau1", commutative = False)
In [7]: evaluate_pauli_product(I \* Pauli(1) \* Pauli(2) \* tau1 \* Pauli(1) \* Pauli(2))
Out[7]: -sigma3 \* sigma1 \* sigma2 \* tau1

With the fixes in this commit, the output of the above is instead:

1.
In [1]: from sympy.physics.paulialgebra import Pauli, evaluate_pauli_product
In [2]: evaluate_pauli_product(1)
Out[2]: 1

2.
In [3]: from sympy import I
In [4]: evaluate_pauli_product(I \* Pauli(1) \* Pauli(2) \* Pauli(1) \* Pauli(2))
Out[4]: -I

3.
In [5]: from sympy import symbols
In [6]: tau1 = symbols("tau1", commutative = False)
In [7]: evaluate_pauli_product(I \* Pauli(1) \* Pauli(2) \* tau1 \* Pauli(1) \* Pauli(2))
Out[7]: -I \* sigma3 \* tau1 \* sigma3

Commit:
paulialgebra: Improved evaluate_pauli_product

evaluate_pauli_product had several limitations before:

1. Acting on arguments that did not have the as_coeff_mul() gave errors.
2. The functions purpose seems to be as a work around for the XFAIL/FIXME that
still remains in the code. But did not produce a satisfying end result.
3. It should respect commutative rules.

This meant that, for example:

1.
>>> from sympy.physics.paulialgebra import Pauli, evaluate_pauli_product
>>> evaluate_pauli_product(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.4/site-packages/sympy/physics/paulialgebra.py",
line 157, in evaluate_pauli_product
    tmp = arg.as_coeff_mul()
AttributeError: 'int' object has no attribute 'as_coeff_mul'

2.
>>> from sympy import I
>>> evaluate_pauli_product(I*Pauli(1)*Pauli(2)*Pauli(1)*Pauli(2))
-sigma3*sigma1*sigma2

3.
>>> from sympy import symbols
>>> tau1 = symbols("tau1", commutative = False)
>>> evaluate_pauli_product(I*Pauli(1)*Pauli(2)*tau1*Pauli(1)*Pauli(2))
-sigma3*sigma1*sigma2*tau1

With the fixes in this commit, the output of the above is instead:

1.
>>> from sympy.physics.paulialgebra import Pauli, evaluate_pauli_product
>>> evaluate_pauli_product(1)
1

2.
>>> from sympy import I
>>> evaluate_pauli_product(I*Pauli(1)*Pauli(2)*Pauli(1)*Pauli(2))
-I

3.
>>> from sympy import symbols
>>> tau1 = symbols("tau1", commutative = False)
>>> evaluate_pauli_product(I*Pauli(1)*Pauli(2)*tau1*Pauli(1)*Pauli(2))
-I*sigma3*tau1*sigma3

bjodah commented:
@johanbluecreek thank you for this contribution. Would you mind adding tests (with comments) for this enhanced behavior in `sympy/physics/tests/test_paulialgebra.py`?

Also pining @alxspopov in case he wants to review this PR.

johanbluecreek commented:
@bjodah Certainly.

Commit:
paulialgebra: Updated evaluate_pauli_product testing

As requested, the test_paulialgebra.py has been updated to include the new
functionality included in Pull Request #11454.

A new test was created to separate tests of Pauli from tests of
evaluate_pauli_product.

moorepants commented:
I don't know what Pauli algebra is, but this change looks looks from from the point of view that it passes existing tests and the new ones. I'm merging as is.
